git/odb/history_rewriter.go

```diff
@@ -0,0 +1,271 @@
+package odb
+
+import (
+	"encoding/hex"
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/git-lfs/git-lfs/git"
+)
+
+// HistoryRewriter allows rewriting topologically equivalent Git histories
+// between two revisions.
+type HistoryRewriter struct {
+	// mu guards entries and commits (see below)
+	mu *sync.Mutex
+	// entries is a mapping of old tree entries to new (rewritten) ones.
+	// Since TreeEntry contains a []byte (and is therefore not a key-able
+	// type), a unique TreeEntry -> string function is used for map keys.
+	entries map[string]*TreeEntry
+	// commits is a mapping of old commit SHAs to new ones, where the ASCII
+	// hex encoding of the SHA1 values are used as map keys.
+	commits map[string][]byte
+	// db is the *ObjectDatabase from which blobs, commits, and trees are
+	// loaded from.
+	db *ObjectDatabase
+}
+
+// RewriteOptions is an options type given to the Rewrite() function.
+type RewriteOptions struct {
+	// Left is the starting commit.
+	Left string
+	// Right is the ending commit.
+	Right string
+
+	// BlobFn specifies a function to rewrite blobs.
+	//
+	// It is called once per unique, unchanged path. That is to say, if
+	// a/foo and a/bar contain identical contents, the BlobFn will be called
+	// twice: once for a/foo and once for a/bar, but no more on each blob
+	// for subsequent revisions, so long as each entry remains unchanged.
+	BlobFn BlobRewriteFn
+}
+
+// BlobRewriteFn is a mapping function that takes a given blob and returns a
+// new, modified blob. If it returns an error, the new blob will not be written
+// and instead the error will be returned from the Rewrite() function.
+//
+// Invocations of an instance of BlobRewriteFn are not expected to store the
+// returned blobs in the *git/odb.ObjectDatabase.
+//
+// The path argument is given to be an absolute path to the tree entry being
+// rewritten, where the repository root is the root of the path given. For
+// instance, a file "b.txt" in directory "dir" would be given as "dir/b.txt",
+// where as a file "a.txt" in the root would be given as "a.txt".
```

Comment:
I think this is a future point to optimize. We shouldn't need to re-process identical blobs multiple times. It looks like there are a couple results of this callback:

1. File path matches `.gitattributes`, and the big file is converted to an LFS pointer.
2. File path does not match `.gitattributes`, but the git blob does exist somewhere else in a path matching `.gitattributes`.
3. No change needed, just return the blob.

I suppose the caller can just implement a callback that saves blob contents or something. Is that the intent? That could make sense given the various ways `HistoryRewriter` can be used.